sdk/python/kfp/compiler/_data_passing_rewriter.py

```diff
@@ -0,0 +1,387 @@
+                        input_name = parts[2]
+                        inputs_directly_consumed_as_parameters.add((template_name, input_name))
+                    else:
+                        raise AssertionError
+                elif placeholder_type == 'tasks':
+                    upstream_task_name = argument_placeholder_parts[1]
+                    assert argument_placeholder_parts[2] == 'outputs'
+                    assert argument_placeholder_parts[3] == 'parameters'
+                    upstream_output_name = argument_placeholder_parts[4]
+                    upstream_template_name = task_name_to_template_name[upstream_task_name]
+                    outputs_directly_consumed_as_parameters.add((upstream_template_name, upstream_output_name))
+                elif placeholder_type == 'workflow' or placeholder_type == 'pod':
+                    pass
+                elif placeholder_type == 'item':
+                    raise AssertionError('The "{{item}}" placeholder is not expected outside task arguments.')
+                else:
+                    raise AssertionError('Unexpected placeholder type "{}".'.format(placeholder_type))
+
+    # Searching for parameter input consumers in container and resource templates
+    for template in container_templates + resource_templates:
+        # Hack for volumes: Prior to Argo 2.3.0, it was only possible to add volumes globally on the workflow level instead of at template level.
+        # When VolumeOps were added, the generated structures became hacky since the workflow-level volumes cound now contain template-level input references.
+        # Here we virtually add volumes back to the template to analyze the input references.
+        # To properly fix this the compiler should move volumes to templates.
+        # TODO: Fix the compiler and remove this hack
+        if 'container' in template:
+            template = copy.deepcopy(template)
+            template_volumes = template.setdefault('volumes', [])
+            for volume_mount in template['container'].get('volumeMounts', []):
+                volume_name = volume_mount['name']
+                if volume_name in volume_map:
+                    template_volumes.append(volume_map[volume_name])
+        # End hack
+
+        template_name = template['name']
+        placeholders = extract_all_placeholders(template)
+        for placeholder in placeholders:
+            parts = placeholder.split('.')
+            placeholder_type = parts[0]
+            if placeholder_type == 'workflow' or placeholder_type == 'pod':
+                pass
+            elif placeholder_type == 'inputs':
+                if parts[1] == 'parameters':
+                    input_name = parts[2]
+                    inputs_directly_consumed_as_parameters.add((template_name, input_name))
+                elif parts[1] == 'artifacts':
+                    raise AssertionError # Should not happen in container templates
```

Comment:
nit: I think it will be great to add some debug info in the AssertionError, too.